frontend-react/docs/best-practices.md

```diff
@@ -0,0 +1,677 @@
+
+Having an explicit dependency on Context values may be intentional in certain cases (e.g., form states), but generally, this should be used in rare circumstances.  See [the React docs](https://reactjs.org/docs/context.html#before-you-use-context) (and [the beta docs](https://beta.reactjs.org/reference/react/useContext)) for more information and alternate approaches.
+
+**Potential performance issues**
+
+Contextual descendants/consumers re-render whenever a new Context value is computed, and this can bypass explicitly memoized components.  This isn't much of a performance hit in small, leaf-node components, but this could also trigger a re-renders of much more significant and complex ancestor nodes -- effectively causing a re-render of the entire tree.  Preventing this requires much more thought and consideration for stabilized Context values.
+
+See [this article](https://blog.thoughtspile.tech/2021/10/04/react-context-dangers/ ) for more information.
+
+#### Try to use effects only when synchronizing with external state or the DOM
+
+> Why?
+>
+> Effects are run after a single render loop so maintaining internal state in effects means that any meaningful change will result in at least two renders: one to flush the initial state to the DOM and then another that immediately follows the flushed DOM.  For internal state, there are usually better solutions like updating in event-handlers or event subscriptions -- or just infer values without using state at all.
+
+```tsx
+function LoadingComponent({ isReady }: { isReady: boolean }) {
+    // bad - no need to use an effect because the state will update _after_ the second render
+    const [isLoading, setIsLoading] = useState(() => !isReady);
+    useEffect(() => {
+        setIsLoading(!isReady);
+    }, [isReady]);
+
+    // good - just infer the value during same render
+    const isLoading = !isReady;
+}
+```  
+
+See [the React docs](https://beta.reactjs.org/learn/you-might-not-need-an-effect) and [this YouTube video](https://www.youtube.com/watch?v=HPoC-k7Rxwo) for more information.
+
+## CSS
+
+### Use classes for low specificity
+
+```scss
+// bad - high specificity (1000)
+<div style="background-color: orange;"></div>
+
+// bad - high specificity (0100)
+#whatever {
+    background-color: yellow;
+}
+
+// bad - lower specificity (0001) than class selector but tag selectors are vague
+div {
+    background-color: blue;
+}
+
+// bad - avoid !importants as much as possible since they ignore specificity
+div {
+    background-color: purple !important;
+}
+
+// good - low specificity (0010) and allows for class composition
+.base-rule {
+    background-color: red;
+}
+
+// good - low specificity (0020) and composed
+.base-rule.variant-rule {
+    background-color: pink;
+}
+```
+
+> Why?
+>
+> Low-specificity selectors in CSS rules makes them more easily overridden if a specific visual variant or ad hoc style needs to be applied.
+
+---
+
```

Comment:
A small section on `_uswds_overrides.scss` and best practices (classes + element selectors of equal specificty > `!important`) and that, at least for now, component-specific CSS should live in: `_components.scss` and wrapped in a css class describing the component. I was doing:

```
// MessageReceivers.tsx
.message-receivers {
  .big-font {
    font-size: 36px;
  }
}

// End -- MessageReceivers.tsx
```

To indicate which component the CSS applied to in my previous work, but open to suggestions!

Reply:
Sure, definitely can do!